### DISCUSSION TITLE
Wiki integrations/plugin

### DISCUSSION THREAD

**Discussion by demiler** [netbox-community/netbox#13345] [Q&A] [answered]
Hi, is it possible to have built-in wiki inside of the netbox? Currently we use redmine's wiki but we are planning on replacing it. Netbox seems like a great place to have technical documentation regarding various configurations, "how-to"s, explained design dissensions, etc. Maybe even simple issue tracker like: "device A not turning on", "ip C is no longer in use".

Maybe I'm trying to make netbox not what it is suppose to be, let me know if I should use other tools for these tasks, I'm really really new to netbox.

I've tried to find any related plugins but with no success

**Comment by mtinberg** [comment, marked as answer]
I can't speak for the Netbox team, but there are plenty of great Wikis, Knowledgebases and Content Management Systems out there and this is a little far from an equipment and config tracking database, I'm not sure that the Netbox dev team could add something unique that would make re-implementing a wiki worth the effort.

That said, the comment feature used in many​ models accepts Markdown which allows for rich text and static links which could point to other sources of documentation.  The Custom URL feature can also create templated links, which could also be used to link to other tools such as a Wiki and include properties of the record in the URL (say linking to a specific doc for each device role).

If you search the netbox website or github you can also find a lot of plugins, including some for document management that add the ability to upload files, to say capture floorplans and as-builts for Sites within netbox rather than just linking to an external tool/CMS.


—

Mark Tinberg ***@***.***>

Division of Information Technology-Network Services

University of Wisconsin-Madison


________________________________
From: John ***@***.***>
Sent: Wednesday, August 2, 2023 6:13 AM
To: netbox-community/netbox ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [netbox-community/netbox] Wiki integrations/plugin (Discussion #13345)


Hi, is it possible to have built-in wiki inside of the netbox? Currently we use redmine's wiki but we are planning on replacing it. Netbox seems like a great place to have technical documentation regarding various configurations, "how-to"s, explained design dissensions, etc. Maybe even simple issue tracker like: "device A not turning on", "ip C is no longer in use".

Maybe I'm trying to make netbox not what it is suppose to be, let me know if I should use other tools for these tasks, I'm really really new to netbox.

I've tried to find any related plugins but with no success

**Comment by candlerb** [comment]
I would use a separate wiki rather than a plugin. As Netbox changes over time, plugins which used to work but are unmaintained may no longer work.  So the fewer third-party plugins you rely on, the better (unless you have the programming skills in-house to maintain them yourself, should it become necessary).

**Reply by demiler** [reply]
Yes, you are probably right, nothing worse than losing all the wiki knowledge base after an update or getting stuck with an old version

**Comment by clint-hotaisle** [comment]
I'm looking for a similar feature. The reason I'd like it in Netbox is I'd like the ability to template data in netbox into my wiki. Like embedding the SVG for a rack diagram (which is really difficult on a different wiki because of browser cross-origin issues).

Ideally it'd support some extended markdown that would allow embedding a custom table view with data from netbox as well.

My end goal is to write a markdown file with adds some contextual information along side a custom view of data in netbox.

Right now if I want to write documentation like this, I need to export the data from netbox which inevitably becomes out of date when the authoritative data in netbox changes.

**Reply by mtinberg** [reply]
I dunno, can the wiki pull from external data sources with an authenticated URL?  You could use an export template to generate HTML for the wiki to embed/consume, proxying through the wiki rather than having the client request the data into an iframe or something.

**Reply by clint-hotaisle** [reply]
This would make writing documentation a pretty painful two-step process of creating an export template in netbox, then creating a wiki page that uses it in another system. The more painful it is to document, the less likely people are to do it. If you want to add another column to a table in a wiki entry, you have to go back and edit the export template instead of editing the wiki page you're on.

I really don't think this is  out of scope. Netbox wants to the the authoritative data source. Notes and documentation should be part of that. Otherwise the data in your documentation and the data in netbox will drift, and people will wonder which is really authoritative.

**Reply by mtinberg** [reply]
I think one of us is misunderstanding the intent here, if you are embedding data from Netbox into a wiki page then you need a mechanism for data to get from netbox to be in-lined with the wiki page, and if you want to edit the source of the netbox data you need to do so in netbox, so either the wiki needs to reach in to netbox (using REST, Export Templates, etc.) or the client needs to embed the content using an iframe, which can run afoul of CORS.

so for example I tied this to virtual-machines

```jinja
{% for vm in queryset %}
{{ vm.name }}</br>
<p>
{{ vm.comments }}
</p>
{% endfor %}
```

then I can export over the API

```shell
$ /usr/local/bin/netbox-query.sh '/virtualization/virtual-machines/?platform_id=123&has_primary_ip=True&export=htmltest' test.html
```

**Reply by clint-hotaisle** [reply]
It's not so much about the source of the data, but the formatting of it. An export template controls what data is selected and displayed. So if I want to update my wiki entry, and I realize there is another piece of data in netbox that would be useful to display on the wiki page, I now also have to go to netbox and edit the export template. So Netbox is the authoratitve source of the data, and a source of part of the formatting, while the wiki is the source of the rest of the formatting. This is pretty confusing.

Netbox is already so close to what is needed with export templates and configuration templates. We just need one more, documentation templates which allow writing jinja templated markdown.

**Reply by mtinberg** [reply]
yeah, I see your point but I'm not sure what a better way would look like, you've got to convert database information to html templates at _some_ point, so I don't see how you avoid having formatting authority split between netbox and the wiki, similar to embedding an outlook or gmail calendar in a wiki page.  I figured if you used the Comment field primarily you could take advantage of the Markdown in Netbox, or at least make some generic embeds that don't need formatting updates very often.  

And to be clear, Export Templates still support all the various query parameters for filtering output, so a lot can be done in the wiki at the point of embedding with generic templates.

Good luck!
